DragaliaAPI/Controllers/Dragalia/FortController.cs

```diff
@@ -43,13 +70,695 @@ await this.fortRepository.GetBuilds(this.DeviceAccountId).ToListAsync()
+
+        // Add carpenter
+        fortDetail.carpenter_num++;
+        await this.fortRepository.UpdateFortCarpenterNum(
+            this.DeviceAccountId, 
+            fortDetail.carpenter_num
+        );
+
+        UpdateDataList updateDataList = this.updateDataService.GetUpdateDataList(
+            this.DeviceAccountId
+        );
+
+        await this.fortRepository.SaveChangesAsync();
+
+        FortAddCarpenterData data =
+            new()
+            {
+                result = 1,
+                fort_detail = fortDetail,
+                update_data_list = updateDataList
+            };
+        return this.Ok(data);
+    }
+
+    [HttpPost("build_at_once")]
+    public async Task<DragaliaResult> BuildAtOnce(FortBuildAtOnceRequest request)
+    {
+        DbPlayerUserData userData = await this.userDataRepository
+            .GetUserData(this.DeviceAccountId)
+            .FirstAsync();
+        FortDetail fortDetail = this.fortRepository.Details.Select(mapper.Map<FortDetail>).First();
+        FortBonusList bonusList = await bonusService.GetBonusList();
+
+        // Get building
+        DbFortBuild build = await this.fortRepository.GetBuilding(
+            this.DeviceAccountId, (long)request.build_id
+        );
+
+        PaymentTypes paymentType = (PaymentTypes)request.payment_type;
+        int paymentHeld = GetUpgradePaymentHeld(userData, paymentType);
+        int paymentCost = GetUpgradePaymentCost(
+            paymentType,
+            build.BuildStartDate,
+            build.BuildEndDate
+        );
+
+        if (paymentHeld < paymentCost)
+        {
+            throw new DragaliaException(
+                ResultCode.FortLevelupIncomplete,
+                $"User did not have enough {paymentType}."
+            );
+        }
+
+        ConsumePaymentCost(userData, paymentType, paymentCost);
+
+        // Update build
+        build.BuildStartDate = DateTimeOffset.UnixEpoch;
+        build.BuildEndDate = DateTimeOffset.UnixEpoch;
+        this.fortRepository.UpdateBuild(build);
+
+        // Update carpenter usage
+        await DecrementCarpenterUsage(fortDetail);
+
+        UpdateDataList updateDataList = this.updateDataService.GetUpdateDataList(
+            this.DeviceAccountId
+        );
+
+        await this.fortRepository.SaveChangesAsync();
+
+        FortBuildAtOnceData data =
+            new()
+            {
+                result = 1,
+                build_id = request.build_id,
+                fort_bonus_list = bonusList,
+                production_rp = StubData.ProductionRp,
+                production_st = StubData.ProductionSt,
+                production_df = StubData.ProductionDf,
+                fort_detail = fortDetail,
+                update_data_list = updateDataList,
+            };
+        return this.Ok(data);
+    }
+
+    [HttpPost("build_cancel")]
+    public async Task<DragaliaResult> BuildCancel(FortBuildCancelRequest request)
```

Comment:
Could you look at extracting the shared logic between build_* and levelup_* endpoints into a service?